<a href="https://colab.research.google.com/github/zwen000/CS448_Assignment_3/blob/main/a3_sample_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

\(\^Be sure to update this button to point to your notebook instead of the sample notebook\)

In [1]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import datasets
from sklearn import svm
from sklearn.linear_model import LinearRegression, BayesianRidge,LogisticRegression
from sklearn.model_selection import cross_val_score,ShuffleSplit,cross_val_predict
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [22]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

df = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
df.head(15)
# df.loc[i][["Temperature °C"]]

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
Temperature °C    1000 non-null int64
Mols KCL          1000 non-null int64
Size nm^3         1000 non-null float64
dtypes: float64(1), int64(2)
memory usage: 23.5 KB


In [24]:
df.describe()
df.size

3000

## Part 2. Splitting the dataset

In [25]:
# Take the pandas dataset and split it into our features (X) and label (y)
features = df[["Temperature °C", "Mols KCL"]]
label = df[["Size nm^3"]]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.1, shuffle=False)
X_train, X_test, y_train, y_test

(     Temperature °C  Mols KCL
 0               469       647
 1               403       694
 2               302       975
 3               779       916
 4               901        18
 5               545       637
 6               660       519
 7               143       869
 8                89       461
 9               294       776
 10              991       117
 11              307       781
 12              206        70
 13              437       599
 14              566        75
 15               62       293
 16              331       781
 17               35       149
 18              905        76
 19              227       864
 20              913        66
 21              709       332
 22              626       903
 23               38       163
 24              849       243
 25              899         5
 26              539       725
 27              230       665
 28              396       846
 29              132       777
 ..              ...       ...
 870    

## Part 3. Perform a Linear Regression

In [26]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_point = np.array([[123, 456]])
y_prediction = model.predict(sample_point)
print("Sample datapoint X: [[123, 456]], Prediction of  y: ", y_prediction )

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print("Score of the model: ", score)
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
intercept = model.intercept_
coef = model.coef_

print ("Intercept: ", intercept)
print ("Coefficents", coef)


Sample datapoint X: [[123, 456]], Prediction of  y:  [[159600.27372536]]
Score of the model:  0.8868651637773253
Intercept:  [-416783.91800716]
Coefficents [[ 879.90717015 1026.65703904]]


# Report
Score = 0.8868651637773253

The score is the coefficient of prediction, it represents the model's capability in predicting the correct label. The best possible score is 1.0, a model with a score closer to 1.0 predicts result closer to the correct label.
Our model has score 0.8868651637773253, very close to 1.0, that is suggesting that our model is fitted pretty well to the existing data, our predicted results will show relatively high accuracy.


# Hypothesis Equation
$
h(x) = \theta_0 + \theta_1 * x_1 + \theta_2 * x_2\\
h(x) = -416783.91800716 + 879.90717015 * x_1 + 1026.65703904 * x_2
$

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [27]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
my_cv = ShuffleSplit(n_splits=10, test_size=0.1, random_state=0)
scores = cross_val_score(model, features, label, cv=my_cv)

print(scores)
print(scores.mean(), scores.std())
# Report on their finding and their significance

[0.87616468 0.86951566 0.83708494 0.86963943 0.84945355 0.86236913
 0.82467112 0.85236386 0.8648058  0.76555589]
0.8471624047034034 0.03112537713964668


The function divided the original dataset and repeat experiment model fitting on the dataset 10 times, with different ways of shuffling each time. Finally, it returned scores for each experiment.

 The scores are rather close to the score of our model from part 3,this shows that our model is fitted well, even shuffling the dataset in different way doesn't improve the performance for our model

## Part 5. Using Polynomial Regression

In [13]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_train = poly.fit_transform(X_train)
X_test = poly.fit_transform(X_test)

model = BayesianRidge()
model.fit(X_train, y_train.values.ravel())


BayesianRidge(alpha_1=1e-06, alpha_2=1e-06, compute_score=False, copy_X=True,
              fit_intercept=True, lambda_1=1e-06, lambda_2=1e-06, n_iter=300,
              normalize=False, tol=0.001, verbose=False)

In [14]:
print("Score: ", model.score(X_test, y_test))
intercept = model.intercept_
coef = model.coef_

print ("Intercept: ", intercept)
print ("Coefficents", coef)


Score:  1.0
Intercept:  2.133921952918172e-05
Coefficents [ 0.00000000e+00  1.20000000e+01 -1.35692972e-07  1.01003650e-11
  2.00000000e+00  2.85714287e-02]


In [15]:
predict_y = model.predict(X_test)
predict_y

array([4.76623457e+05, 5.54434600e+05, 1.07736803e+06, 8.18464114e+05,
       3.18040000e+04, 6.70937114e+05, 5.13347114e+05, 5.20717457e+05,
       3.68407143e+04, 2.26740429e+05, 6.72088457e+05, 2.15877600e+05,
       5.37476829e+05, 4.13732829e+05, 7.58377114e+05, 1.07312314e+05,
       1.43706126e+06, 3.58148457e+05, 1.60323531e+06, 3.46296114e+05,
       1.06297831e+06, 8.96234571e+04, 6.16569857e+05, 3.87182257e+05,
       4.80946400e+05, 9.76260286e+04, 8.28280000e+04, 4.73076600e+05,
       9.54178829e+05, 6.61005000e+05, 1.14400011e+06, 6.12330286e+04,
       9.29079314e+05, 1.03529243e+06, 5.02000000e+04, 1.70761660e+06,
       2.16700286e+04, 7.74480114e+05, 3.55214714e+05, 7.38879429e+05,
       2.51469829e+05, 2.53718829e+05, 5.21228571e+04, 2.30945716e+03,
       4.22685029e+05, 2.00866257e+05, 3.47416457e+05, 1.50313257e+05,
       1.72224626e+06, 1.65311430e+03, 4.81235314e+05, 1.37193257e+05,
       4.26357257e+05, 1.32183143e+06, 6.74216000e+04, 7.26984429e+05,
      

In [16]:
print(y_test)

        Size nm^3
900  4.766235e+05
901  5.544346e+05
902  1.077368e+06
903  8.184641e+05
904  3.180400e+04
905  6.709371e+05
906  5.133471e+05
907  5.207175e+05
908  3.684071e+04
909  2.267404e+05
910  6.720885e+05
911  2.158776e+05
912  5.374768e+05
913  4.137328e+05
914  7.583771e+05
915  1.073123e+05
916  1.437061e+06
917  3.581485e+05
918  1.603235e+06
919  3.462961e+05
920  1.062978e+06
921  8.962346e+04
922  6.165699e+05
923  3.871823e+05
924  4.809464e+05
925  9.762603e+04
926  8.282800e+04
927  4.730766e+05
928  9.541788e+05
929  6.610050e+05
..            ...
970  2.786154e+05
971  3.176299e+05
972  4.598003e+05
973  1.286107e+05
974  1.140281e+05
975  1.051815e+06
976  6.321883e+05
977  2.407453e+05
978  9.569281e+05
979  5.982961e+05
980  1.642144e+05
981  1.422398e+06
982  1.964650e+05
983  7.396408e+05
984  1.279600e+06
985  1.968257e+03
986  7.975773e+05
987  4.155634e+05
988  4.469683e+04
989  1.302394e+05
990  9.089383e+04
991  1.331614e+06
992  4.871523e+05
993  1.705

In [17]:
# Report on the metrics and output the resultant equation as you did in Part 3.

# Hypothesis Equation
$
h(x) = 2.133921952918172\mathrm{e}{-05} + 0.00000000 + 1.20000000\mathrm{e}{+01} * a - 1.35692972\mathrm{e}{-07} * b + 1.01003650\mathrm{e}{-11} * a^2 + 2.00000000 * ab + 2.85714287\mathrm{e}{-02} * b^2
$

# Report
I used BayesianRidge()in part 5,the model's score is 1.0,the highest possible score for a model, the model can predict the correct result for sure.